### PR TITLE
machines: Introduce Storage Pools listing

### PR DESCRIPTION
**Pull request by KKoukiou**

### CODE REVIEW

**Comment by KKoukiou** [comment]
Here is a very naive implementation of listing the storage pools. It's still in progress, but I wanted some design feedback.

@garrett it's time to decide where we need to place the storage pools information.

For now it's just a static table, in the future I will make the storage pool rows expandable, and there will be multiple tabs, like we do for the VMs listing. In these tab, I 'll store information about the storage pool itself and also the storage volumes it contains.

One more thing I don't really like is how we mix session/system connection resources, and I would like that we handle that somehow.

storage-pools-lsiting
Note: I know the connection column is not aligned with the Vms connection column, I couldn't figure out the problem yet. But again this is just a draft.

**Comment by andreasn** [comment]
I get an "Oops" in Cockpit when I run this.
Here is the console output:
```
TypeError: _providerActions.getStorageVolumes is not a function[Learn More] 2 machines.js:124912:32
	GET_STORAGE_POOL/</< https://192.168.1.207:9090/cockpit/$4d703d014c3eb5a0245d59408745485761406a9d1f7140f991f47eb8b19a032e/machines/machines.js:124912:32
	T https://192.168.1.207:9090/cockpit/$4d703d014c3eb5a0245d59408745485761406a9d1f7140f991f47eb8b19a032e/base1/cockpit.js:879:31
	D/< https://192.168.1.207:9090/cockpit/$4d703d014c3eb5a0245d59408745485761406a9d1f7140f991f47eb8b19a032e/base1/cockpit.js:891:17
	j https://192.168.1.207:9090/cockpit/$4d703d014c3eb5a0245d59408745485761406a9d1f7140f991f47eb8b19a032e/base1/cockpit.js:808:17
```

**Comment by KKoukiou** [comment]
> I get an "Oops" in Cockpit when I run this.
> Here is the console output:
> 
> ```
> TypeError: _providerActions.getStorageVolumes is not a function[Learn More] 2 machines.js:124912:32
> 	GET_STORAGE_POOL/</< https://192.168.1.207:9090/cockpit/$4d703d014c3eb5a0245d59408745485761406a9d1f7140f991f47eb8b19a032e/machines/machines.js:124912:32
> 	T https://192.168.1.207:9090/cockpit/$4d703d014c3eb5a0245d59408745485761406a9d1f7140f991f47eb8b19a032e/base1/cockpit.js:879:31
> 	D/< https://192.168.1.207:9090/cockpit/$4d703d014c3eb5a0245d59408745485761406a9d1f7140f991f47eb8b19a032e/base1/cockpit.js:891:17
> 	j https://192.168.1.207:9090/cockpit/$4d703d014c3eb5a0245d59408745485761406a9d1f7140f991f47eb8b19a032e/base1/cockpit.js:808:17
> ```

@andreasn  it's because  you don't have libvirt-dbus package installed.
This PR is a draft and I implemented the change only for the part of the code using libvirt-dbus. I 'll have to fix it before merging for virsh as well, but please install libvirt-dbus package if you want to test it to give design feedback for now :)

**Comment by andreasn** [comment]
Not final, mostly for getting some ideas out of my system.
@garrett is working on a more final version

The critical part in this one is that it focuses on answering "how much storage do I have left", and "what can I do to free up storage".

mockup

**Comment by garrett** [comment]
Echoing here what I said earlier in IRC and the meeting: I really like the way the storage pools list has the inline storage used bar in @andreasn's mockups above

My other mockups in progress weren't really compatible with this list, and I think the card dashboard idea could work well — but we need to modify the cards to adapt them into our layout. Therefore, I decided to play around with ideas for the cards @ https://garrett.github.io/cockpit-mockweb/cards/

I'll take what I have there and integrate it into a merged mockup between that and what Andreas posted above.

What I think needs changing in the above mockup:

1. We should have storage pools and virtual networks as cards, but not the virtual machines, as that is below in the list. (This is part of the reason why I worked on the card modifications, so that 1 - 2 cards could work, and would be displayed nicely.)

2. That's about it, really. Looks great!

Additional things to consider:

1. What happens to VMs when storage they use happens to be removed from the storage pool page?

2. Should we show if storage is configured to be used by a VM? (Yes.)

3. Storage pool interaction when creating / modifying a VM. (Creating a disk in a specific pool, selecting an existing disk from a pool, etc.) — probably a separate PR.

4. What statistics should the cards show? I think it should be:
   1. Storage Pools:
      - number of pools configured
      - ~used space~ (probably not relevant in the card)
      - available space
      - ~total space~ (also probably not relevant in the card)
      - warnings/errors (if disks are offline or do not have enough space)
    2. Virtual Disks:
       - Not really relevant at this time, except for contextual design: probably number of networks and warnings/errors

**Comment by garrett** [comment]
On second thought, I wonder if we should even have all the additional statistics. We could probably just get away with the number and warnings/errors — especially for this PR. (A follow-up could add more info, if we decide that it's a good idea.)

**Comment by garrett** [comment]
Modified @andreasn's mockup with grid-based "hybrid" card layout.

## Static mockup

storage pools mockup

The big change was adapting the cards in a simple manner. 

When there's only 1 card, like there will be after this PR, it takes up ½ the space automatically instead of stretching across the width. This requires the new CSS in the card mockup and the different layout wrapper with the new classname(s).

## HTML mockup

https://garrett.github.io/cockpit-mockweb/machines/
(Click on storage pools.)

**Comment by garrett** [comment]
@KKoukiou The wrapper HTML is just the following:

```html
<div class="cards-pf grid-cards-ct cards-ct-hybrid">
</div>
```

There are no bootstrap containers or rows or cols or any of that. `cards-pf` provides the grey background — we can leave this out for a white background and everything should still work fine too.

The cards inside can be rendered with PatternFly-React and will look kinda like this when rendered as HTML:

```html
  <div class="card-pf card-pf-accented card-pf-aggregate-status">
    <h3 class="card-pf-title">
      <a href="pools"><span class="pficon pficon-server"></span><span class="card-pf-aggregate-status-count">20</span>
        Storage Pools
      </a>
    </h3>
    <div class="card-pf-body">
      <p class="card-pf-aggregate-status-notifications">
        <span class="card-pf-aggregate-status-notification"><a href="#"><span class="pficon pficon-warning-triangle-o"></span>1</a></span>
      </p>
    </div>
  </div>
```

It should do "the right thing", provided the new CSS is active. We should try to get the CSS in a non-machines part of Cockpit.

**Comment by garrett** [comment]
I _think_ we can put custom card CSS in something `pkg/lib/cards.less` and then include/require it (depending on what pulls in LESS) in the machines CSS.

At some point in time, we probably want to make a custom library CSS file (compiled from other files, I guess) which puts ALL the common CSS shared across Cockpit into one file. But until then, I think this is correct?

@martinpitt, @mvollmer: Is the above information right? Please correct if not. Thanks!

**Comment by martinpitt** [comment]
@garrett: Yes, that should work.

**Comment by andreasn** [comment]
The progress class from Patternfly has a margin-bottom of 20px set. This needs to be overridden in the cases where the progressbar is inside an expander.

screenshot from 2018-11-06 12-37-09

**Comment by garrett** [comment]
@andreasn: Yep! :+1: Took care of it (among tons of little things) in cad4007.

**Comment by andreasn** [comment]
I couldn't figure out the right css for it yet, but I think it would be a good idea to decrease the padding a bit on the items under Storage Volumes. It scales very badly right now. Might be worth introducing a max-height limit also, and have it start scrolling at some point.

This is how it looks now:
screenshot from 2018-11-06 16-16-39

File: pkg/machines/components/storagePools/storagePool.es6 [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    ];
+
+    const overviewTabName = (
+        <div id={`${storagePoolId(storagePool.name)}-${storagePool.connectionName}-overview`}>
+            {_("Overview")}
+        </div>
+    );
+    const storageVolsTabName = (
+        <div id={`${storagePoolId(storagePool.name)}-${storagePool.connectionName}-storage-volumes`}>
+            {_("Storage Volumes")}
```
Comment (andreasn) [inline pkg/machines/components/storagePools/storagePool.es6]: A bit of nitpicking, but for simplicity, maybe we should call this just _Volumes_
The context itself makes it pretty clear these are storage volumes and not any other kind of volumes. It's also the string used in virt-manager today.

**Comment by garrett** [comment]
@andreasn I'll make a compact style for the list and introduce it here.

**Comment by andreasn** [comment]
>  I'll make a compact style for the list and introduce it here.

Awesome, thanks!

**Comment by andreasn** [comment]
Piling on another observation. Not really a blocker, and we can split it out as a followup PR if needed

It seems it's listing the full path under _Storage Volumes_, repeating the info in the _Overview_ tab.
screenshot from 2018-11-06 16-54-46

**Comment by garrett** [comment]
I figured out how to do compact lists (in a generic way, for all of Cockpit). It's in a separate commit and I have another commit to enable it for the volume list.

It still needs some work; putting this off for tomorrow:
- [ ] The expandable divider is improperly offset (not an issue with this particular list at least)
- [ ] Divider lines are too dark, especially in the compact lists
- [ ] Headers should probably be adjusted as well

**Comment by garrett** [comment]
Some lists cause a crash. It appease they are empty. The page goes blank, and I get this error in my console:

```
The above error occurred in the <StoragePoolVolumesTab> component:
    in StoragePoolVolumesTab (created by ListingRow)
    in div (created by ListingRow)
    in td (created by ListingRow)
    in tr (created by ListingRow)
    in tbody (created by ListingRow)
    in ListingRow (created by StoragePool)
    in StoragePool (created by StoragePoolList)
    in table (created by Listing)
    in Listing (created by StoragePoolList)
    in div (created by StoragePoolList)
    in StoragePoolList (created by App)
    in div (created by App)
    in App

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://fb.me/react-error-boundaries to learn more about error boundaries. machines.js:18209:6
logCapturedError
https://sunny:9090/cockpit/@localhost/machines/machines.js:18209:6
logError
https://sunny:9090/cockpit/@localhost/machines/machines.js:18243:6
createRootErrorUpdate/update.callback
https://sunny:9090/cockpit/@localhost/machines/machines.js:18988:6
callCallback
https://sunny:9090/cockpit/@localhost/machines/machines.js:14280:4
commitUpdateEffects
https://sunny:9090/cockpit/@localhost/machines/machines.js:14319:8
commitUpdateQueue
https://sunny:9090/cockpit/@localhost/machines/machines.js:14310:4
commitLifeCycles
https://sunny:9090/cockpit/@localhost/machines/machines.js:18376:12
commitAllLifeCycles
https://sunny:9090/cockpit/@localhost/machines/machines.js:19609:8
callCallback
https://sunny:9090/cockpit/@localhost/machines/machines.js:3235:10
invokeGuardedCallbackDev
https://sunny:9090/cockpit/@localhost/machines/machines.js:3285:8
invokeGuardedCallback
https://sunny:9090/cockpit/@localhost/machines/machines.js:3342:4
commitRoot
https://sunny:9090/cockpit/@localhost/machines/machines.js:19763:8
completeRoot
https://sunny:9090/cockpit/@localhost/machines/machines.js:21155:4
performWorkOnRoot
https://sunny:9090/cockpit/@localhost/machines/machines.js:21083:10
performWork
https://sunny:9090/cockpit/@localhost/machines/machines.js:20987:8
performSyncWork
https://sunny:9090/cockpit/@localhost/machines/machines.js:20959:4
interactiveUpdates$1
https://sunny:9090/cockpit/@localhost/machines/machines.js:21250:8
interactiveUpdates
https://sunny:9090/cockpit/@localhost/machines/machines.js:5303:11
dispatchInteractiveEvent
https://sunny:9090/cockpit/@localhost/machines/machines.js:8002:4
TypeError: storagePool.volumes is undefined
```

File: pkg/machines/components/diskAdd.jsx [hunk 2]
```diff
@@ -310,13 +310,23 @@ export class AddDiskAction extends React.Component {
     render() {
         const { vm, storagePools, provider, dispatch } = this.props;
         const idPrefix = `${this.props.idPrefix}-adddisk`;
+        /*
+         * For now only Storage Pools of dir type are supported.
+         * TODO: Add support for other Storage Pool types.
+         */
+        const filteredStoragePools = storagePools
+                .filter(pool => pool.connectionName == vm.connectionName && pool.active && pool.type == 'dir')
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/components/diskAdd.jsx]: In order to use a volume from storage pools to add it as disk the pools needs to be active. This behavior is not new, it existed before, we just filtered the storage pools when getting then from the API.
See:
https://github.com/cockpit-project/cockpit/blob/c2b09aa041f7c7d41bc4050d5318a1436a44d757/pkg/machines/libvirt-dbus.es6#L468

No we added the tables with the storage pools, so we would like to fetch from the API all storage pools. The disk add component that needs only the active ones, should filter them by itself, which is the change you see here.

File: pkg/machines/components/diskAdd.jsx [hunk 3]
```diff
@@ -310,13 +310,23 @@ export class AddDiskAction extends React.Component {
...
+                .reduce((result, filter) => {
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/components/diskAdd.jsx]: Not sure what you mean by strange name :) ?
I am using the builtin ```Array.prototype.filter()``` to filter the array of the pools.
I am using reduce the transform the array of storage pools from the global state, to the format the diskAdd component expects.

File: pkg/machines/constants/store-action-types.es6 [hunk 4]
```diff
@@ -29,10 +29,10 @@ export const SET_LOGGED_IN_USER = "SET_LOGGED_IN_USER";
 export const SET_PROVIDER = "SET_PROVIDER";
 export const UNDEFINE_VM = "UNDEFINE_VM";
 export const UPDATE_ADD_VM = "UPDATE_ADD_VM";
+export const UPDATE_ADD_STORAGE_POOL = "UPDATE_ADD_STORAGE_POOL";
 export const UPDATE_LIBVIRT_STATE = "UPDATE_LIBVIRT_STATE";
 export const UPDATE_NETWORKS = "UPDATE_NETWORKS";
 export const UPDATE_OS_INFO_LIST = "UPDATE_OS_INFO_LIST";
-export const UPDATE_STORAGE_POOLS = "UPDATE_STORAGE_POOLS";
 export const UPDATE_STORAGE_VOLUMES = "UPDATE STORAGE_VOLUMES";
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/constants/store-action-types.es6]: Oh, sure! I 'll post a patch with this PR.

File: pkg/machines/index.es6 [hunk 5]
```diff
@@ -56,7 +56,7 @@ function detectLibvirtProvider() {
 
 function render() {
     ReactDOM.render(
-        React.createElement(App, {store: store}),
+        <App store={store} />,
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/index.es6]: As we discussed over IRC, es6 to be used to libraries and jsx fo react components. Although there is not really any difference on how they are parsed.
This file was es6 before, I 'll have it in mind for future though :)

File: pkg/machines/helpers.es6 [hunk 6]
```diff
@@ -297,6 +297,10 @@ export function vmId(vmName) {
     return `vm-${vmName}`;
 }
 
+export function storagePoolId(poolName) {
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/helpers.es6]: Right, will fix.

File: pkg/machines/components/diskAdd.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/machines/components/diskAdd.jsx]: Ahh, thanks!

File: pkg/machines/components/diskAdd.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/machines/components/diskAdd.jsx]: I meant the name of the argument of the function given to `reduce`.  What about `reduce((result, pool) => { ... })`?

**Comment by martinpitt** [comment]
Video: https://www.youtube.com/watch?v=bPx3_LEn9WI&feature=youtu.be

**Pull request merged**